<<<<<<< HEAD
# jsx-a11y/aria-proptypes

ðŸ’¼ This rule is enabled in the following configs: â˜‘ï¸ `recommended`, ðŸ”’ `strict`.

<!-- end auto-generated rule header -->

ARIA state and property values must be valid.

=======
# aria-proptypes

ARIA state and property values must be valid.

#### References
1. [Spec](https://www.w3.org/TR/wai-aria/states_and_properties)
2. [AX_ARIA_04](https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_aria_04)

>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
## Rule details

This rule takes no arguments.

### Succeed
```jsx
<!-- Good: the aria-hidden state is of type true/false -->
<span aria-hidden="true">foo</span>
```

### Fail
```jsx
<!-- Bad: the aria-hidden state is of type true/false -->
<span aria-hidden="yes">foo</span>
```

<<<<<<< HEAD
## Accessibility guidelines
- [WCAG 4.1.2](https://www.w3.org/WAI/WCAG21/Understanding/name-role-value)

### Resources
- [ARIA Spec, States and Properties](https://www.w3.org/TR/wai-aria/#states_and_properties)
- [Chrome Audit Rules, AX_ARIA_04](https://github.com/GoogleChrome/accessibility-developer-tools/wiki/Audit-Rules#ax_aria_04)
=======
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
