<<<<<<< HEAD
---
description: 'Disallow awaiting a value that is not a Thenable.'
---

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/await-thenable** for documentation.

A "Thenable" value is an object which has a `then` method, such as a Promise.
The `await` keyword is generally used to retrieve the result of calling a Thenable's `then` method.

If the `await` keyword is used on a value that is not a Thenable, the value is directly resolved immediately.
While doing so is valid JavaScript, it is often a programmer error, such as forgetting to add parenthesis to call a function that returns a Promise.

## Examples

<!--tabs-->

### ❌ Incorrect
=======
# Disallows awaiting a value that is not a Thenable (`await-thenable`)

This rule disallows awaiting a value that is not a "Thenable" (an object which has `then` method, such as a Promise).
While it is valid JavaScript to await a non-`Promise`-like value (it will resolve immediately), this pattern is often a programmer error, such as forgetting to add parenthesis to call a function that returns a Promise.

## Rule Details

Examples of **incorrect** code for this rule:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
await 'value';

const createValue = () => 'value';
await createValue();
```

<<<<<<< HEAD
### ✅ Correct
=======
Examples of **correct** code for this rule:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
await Promise.resolve('value');

const createValue = async () => 'value';
await createValue();
```

## When Not To Use It

If you want to allow code to `await` non-Promise values.
This is generally not preferred, but can sometimes be useful for visual consistency.
<<<<<<< HEAD
=======

## Related to

- TSLint: ['await-promise'](https://palantir.github.io/tslint/rules/await-promise)
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
