<<<<<<< HEAD
---
description: 'Enforce default parameters to be last.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/default-param-last** for documentation.

## Examples
=======
# Enforce default parameters to be last (`default-param-last`)

## Rule Details
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

This rule extends the base [`eslint/default-param-last`](https://eslint.org/docs/rules/default-param-last) rule.
It adds support for optional parameters.

<<<<<<< HEAD
<!--tabs-->

### ‚ùå Incorrect

```ts
/* eslint @typescript-eslint/default-param-last: "error" */
=======
Examples of **incorrect** code for this rule:

```ts
/* eslint @typescript-eslint/default-param-last: ["error"] */
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

function f(a = 0, b: number) {}
function f(a: number, b = 0, c: number) {}
function f(a: number, b?: number, c: number) {}
class Foo {
  constructor(public a = 10, private b: number) {}
}
class Foo {
  constructor(public a?: number, private b: number) {}
}
```

<<<<<<< HEAD
### ‚úÖ Correct

```ts
/* eslint @typescript-eslint/default-param-last: "error" */
=======
Examples of **correct** code for this rule:

```ts
/* eslint @typescript-eslint/default-param-last: ["error"] */
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

function f(a = 0) {}
function f(a: number, b = 0) {}
function f(a: number, b?: number) {}
function f(a: number, b?: number, c = 0) {}
function f(a: number, b = 0, c?: number) {}
class Foo {
  constructor(public a, private b = 0) {}
}
class Foo {
  constructor(public a, private b?: number) {}
}
```
<<<<<<< HEAD
=======

## How to use

```jsonc
{
  // note you must disable the base rule as it can report incorrect errors
  "default-param-last": "off",
  "@typescript-eslint/default-param-last": ["error"]
}
```

## Options

See [`eslint/default-param-last` options](https://eslint.org/docs/rules/default-param-last#options).

<sup>Taken with ‚ù§Ô∏è [from ESLint core](https://github.com/eslint/eslint/blob/master/docs/rules/default-param-last.md)</sup>
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
