<<<<<<< HEAD
---
description: 'Disallow non-null assertions using the `!` postfix operator.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-non-null-assertion** for documentation.

TypeScript's `!` non-null assertion operator asserts to the type system that an expression is non-nullable, as in not `null` or `undefined`.
Using assertions to tell the type system new information is often a sign that code is not fully type-safe.
It's generally better to structure program logic so that TypeScript understands when values may be nullable.

## Examples

<!--tabs-->

### âŒ Incorrect

```ts
interface Example {
  property?: string;
}

declare const example: Example;
const includesBaz = example.property!.includes('baz');
```

### âœ… Correct

```ts
interface Example {
  property?: string;
}

declare const example: Example;
const includesBaz = example.property?.includes('baz') ?? false;
=======
# Disallows non-null assertions using the `!` postfix operator (`no-non-null-assertion`)

## Rule Details

Using non-null assertions cancels the benefits of the strict null-checking mode.

Examples of **incorrect** code for this rule:

```ts
interface Foo {
  bar?: string;
}

const foo: Foo = getFoo();
const includesBaz: boolean = foo.bar!.includes('baz');
```

Examples of **correct** code for this rule:

```ts
interface Foo {
  bar?: string;
}

const foo: Foo = getFoo();
const includesBaz: boolean = foo.bar && foo.bar.includes('baz');
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
```

## When Not To Use It

<<<<<<< HEAD
If your project does not use the `strictNullChecks` compiler option, this rule is likely useless to you.
If your code is often wildly incorrect with respect to strict null-checking, your code may not yet be ready for this rule.
=======
If you don't care about strict null-checking, then you will not need this rule.

## Further Reading

- [`no-non-null-assertion`](https://palantir.github.io/tslint/rules/no-non-null-assertion/) in [TSLint](https://palantir.github.io/tslint/)
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
