<<<<<<< HEAD
---
description: 'Disallow the use of parameter properties in class constructors.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-parameter-properties** for documentation.

:::danger Deprecated

This rule has been deprecated in favour of the equivalent, better named [`parameter-properties`](./parameter-properties.md) rule.
:::
=======
# Disallow the use of parameter properties in class constructors (`no-parameter-properties`)
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

Parameter properties can be confusing to those new to TypeScript as they are less explicit than other ways
of declaring and initializing class members.

<<<<<<< HEAD
## Examples
=======
## Rule Details
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

This rule disallows the use of parameter properties in constructors, forcing the user to explicitly
declare all properties in the class.

## Options

This rule, in its default state, does not require any argument and would completely disallow the use of parameter properties.
If you would like to allow certain types of parameter properties then you may pass an object with the following options:

- `allows`, an array containing one or more of the allowed modifiers. Valid values are:
  - `readonly`, allows **readonly** parameter properties.
  - `private`, allows **private** parameter properties.
  - `protected`, allows **protected** parameter properties.
  - `public`, allows **public** parameter properties.
  - `private readonly`, allows **private readonly** parameter properties.
  - `protected readonly`, allows **protected readonly** parameter properties.
  - `public readonly`, allows **public readonly** parameter properties.

### default

<<<<<<< HEAD
Examples of code for this rule with no options at all:

<!--tabs-->

#### âŒ Incorrect
=======
Examples of **incorrect** code for this rule with no options at all:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(readonly name: string) {}
}

class Foo {
  constructor(private name: string) {}
}

class Foo {
  constructor(protected name: string) {}
}

class Foo {
  constructor(public name: string) {}
}

class Foo {
  constructor(private readonly name: string) {}
}

class Foo {
  constructor(protected readonly name: string) {}
}

class Foo {
  constructor(public readonly name: string) {}
}
```

<<<<<<< HEAD
#### âœ… Correct
=======
Examples of **correct** code for this rule with no options at all:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(name: string) {}
}
```

### readonly

<<<<<<< HEAD
Examples of code for the `{ "allows": ["readonly"] }` options:

<!--tabs-->

#### âŒ Incorrect
=======
Examples of **incorrect** code for the `{ "allows": ["readonly"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(private name: string) {}
}

class Foo {
  constructor(protected name: string) {}
}

class Foo {
  constructor(public name: string) {}
}

class Foo {
  constructor(private readonly name: string) {}
}

class Foo {
  constructor(protected readonly name: string) {}
}

class Foo {
  constructor(public readonly name: string) {}
}
```

<<<<<<< HEAD
#### âœ… Correct
=======
Examples of **correct** code for the `{ "allows": ["readonly"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(name: string) {}
}

class Foo {
  constructor(readonly name: string) {}
}
```

### private

<<<<<<< HEAD
Examples of code for the `{ "allows": ["private"] }` options:

<!--tabs-->

#### âŒ Incorrect
=======
Examples of **incorrect** code for the `{ "allows": ["private"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(readonly name: string) {}
}

class Foo {
  constructor(protected name: string) {}
}

class Foo {
  constructor(public name: string) {}
}

class Foo {
  constructor(private readonly name: string) {}
}

class Foo {
  constructor(protected readonly name: string) {}
}

class Foo {
  constructor(public readonly name: string) {}
}
```

<<<<<<< HEAD
#### âœ… Correct
=======
Examples of **correct** code for the `{ "allows": ["private"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(name: string) {}
}

class Foo {
  constructor(private name: string) {}
}
```

### protected

<<<<<<< HEAD
Examples of code for the `{ "allows": ["protected"] }` options:

<!--tabs-->

#### âŒ Incorrect
=======
Examples of **incorrect** code for the `{ "allows": ["protected"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(readonly name: string) {}
}

class Foo {
  constructor(private name: string) {}
}

class Foo {
  constructor(public name: string) {}
}

class Foo {
  constructor(private readonly name: string) {}
}

class Foo {
  constructor(protected readonly name: string) {}
}

class Foo {
  constructor(public readonly name: string) {}
}
```

<<<<<<< HEAD
#### âœ… Correct
=======
Examples of **correct** code for the `{ "allows": ["protected"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(name: string) {}
}

class Foo {
  constructor(protected name: string) {}
}
```

### public

<<<<<<< HEAD
Examples of code for the `{ "allows": ["public"] }` options:

<!--tabs-->

#### âŒ Incorrect
=======
Examples of **incorrect** code for the `{ "allows": ["public"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(readonly name: string) {}
}

class Foo {
  constructor(private name: string) {}
}

class Foo {
  constructor(protected name: string) {}
}

class Foo {
  constructor(private readonly name: string) {}
}

class Foo {
  constructor(protected readonly name: string) {}
}

class Foo {
  constructor(public readonly name: string) {}
}
```

<<<<<<< HEAD
#### âœ… Correct
=======
Examples of **correct** code for the `{ "allows": ["public"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(name: string) {}
}

class Foo {
  constructor(public name: string) {}
}
```

### private readonly

<<<<<<< HEAD
Examples of code for the `{ "allows": ["private readonly"] }` options:

<!--tabs-->

#### âŒ Incorrect
=======
Examples of **incorrect** code for the `{ "allows": ["private readonly"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(readonly name: string) {}
}

class Foo {
  constructor(private name: string) {}
}

class Foo {
  constructor(protected name: string) {}
}

class Foo {
  constructor(public name: string) {}
}

class Foo {
  constructor(protected readonly name: string) {}
}

class Foo {
  constructor(public readonly name: string) {}
}
```

<<<<<<< HEAD
#### âœ… Correct
=======
Examples of **correct** code for the `{ "allows": ["private readonly"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(name: string) {}
}

class Foo {
  constructor(private readonly name: string) {}
}
```

### protected readonly

<<<<<<< HEAD
Examples of code for the `{ "allows": ["protected readonly"] }` options:

<!--tabs-->

#### âŒ Incorrect
=======
Examples of **incorrect** code for the `{ "allows": ["protected readonly"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(readonly name: string) {}
}

class Foo {
  constructor(private name: string) {}
}

class Foo {
  constructor(protected name: string) {}
}

class Foo {
  constructor(public name: string) {}
}

class Foo {
  constructor(private readonly name: string) {}
}

class Foo {
  constructor(public readonly name: string) {}
}
```

<<<<<<< HEAD
#### âœ… Correct
=======
Examples of **correct** code for the `{ "allows": ["protected readonly"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(name: string) {}
}

class Foo {
  constructor(protected readonly name: string) {}
}
```

### public readonly

<<<<<<< HEAD
Examples of code for the `{ "allows": ["public readonly"] }` options:

<!--tabs-->

#### âŒ Incorrect
=======
Examples of **incorrect** code for the `{ "allows": ["public readonly"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(readonly name: string) {}
}

class Foo {
  constructor(private name: string) {}
}

class Foo {
  constructor(protected name: string) {}
}

class Foo {
  constructor(public name: string) {}
}

class Foo {
  constructor(private readonly name: string) {}
}

class Foo {
  constructor(protected readonly name: string) {}
}
```

<<<<<<< HEAD
#### âœ… Correct
=======
Examples of **correct** code for the `{ "allows": ["public readonly"] }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Foo {
  constructor(name: string) {}
}

class Foo {
  constructor(public readonly name: string) {}
}
```

## When Not To Use It

If you don't care about the using parameter properties in constructors, then you will not need this rule.
<<<<<<< HEAD
=======

## Compatibility

- TSLint: [no-parameter-properties](https://palantir.github.io/tslint/rules/no-parameter-properties/)
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
