<<<<<<< HEAD
---
description: 'Disallow unnecessary namespace qualifiers.'
---

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/no-unnecessary-qualifier** for documentation.

Members of TypeScript enums and namespaces are generally retrieved as qualified property lookups: e.g. `Enum.member`.
However, when accessed within their parent enum or namespace, the qualifier is unnecessary: e.g. just `member` instead of `Enum.member`.
This rule reports when an enum or namespace qualifier is unnecessary.

## Examples

<!--tabs-->

### ❌ Incorrect
=======
# Warns when a namespace qualifier is unnecessary (`no-unnecessary-qualifier`)

## Rule Details

This rule aims to let users know when a namespace or enum qualifier is unnecessary,
whether used for a type or for a value.

Examples of **incorrect** code for this rule:

```ts
namespace A {
  export type B = number;
  const x: A.B = 3;
}
```

```ts
namespace A {
  export const x = 3;
  export const y = A.x;
}
```
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
enum A {
  B,
  C = A.B,
}
```

```ts
namespace A {
<<<<<<< HEAD
  export type B = number;
  const x: A.B = 3;
}
```

### ✅ Correct

```ts
enum A {
  B,
  C = B,
=======
  export namespace B {
    export type T = number;
    const x: A.B.T = 3;
  }
}
```

Examples of **correct** code for this rule:

```ts
namespace X {
  export type T = number;
}

namespace Y {
  export const x: X.T = 3;
}
```

```ts
enum A {
  X,
  Y,
}

enum B {
  Z = A.X,
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
}
```

```ts
<<<<<<< HEAD
namespace A {
  export type B = number;
  const x: B = 3;
=======
namespace X {
  export type T = number;
  namespace Y {
    type T = string;
    const x: X.T = 0;
  }
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
}
```

## When Not To Use It

<<<<<<< HEAD
If you don't care about having unneeded enum or namespace qualifiers, then you don't need to use this rule.
=======
If you don't care about having unneeded namespace or enum qualifiers, then you don't need to use this rule.

## Further Reading

- TSLint: [no-unnecessary-qualifier](https://palantir.github.io/tslint/rules/no-unnecessary-qualifier/)
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
