<<<<<<< HEAD
---
description: 'Disallow calling a value with type `any`.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-unsafe-call** for documentation.

The `any` type in TypeScript is a dangerous "escape hatch" from the type system.
Using `any` disables many type checking rules and is generally best used only as a last resort or when prototyping code.

Despite your best intentions, the `any` type can sometimes leak into your codebase.
Calling an `any`-typed value as a function creates a potential type safety hole and source of bugs in your codebase.

This rule disallows calling any value that is typed as `any`.

## Examples

<!--tabs-->

### âŒ Incorrect
=======
# Disallows calling an any type value (`no-unsafe-call`)

Despite your best intentions, the `any` type can sometimes leak into your codebase.
The arguments to, and return value of calling an `any` typed variable are not checked at all by TypeScript, so it creates a potential safety hole, and source of bugs in your codebase.

## Rule Details

This rule disallows calling any variable that is typed as `any`.

Examples of **incorrect** code for this rule:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
declare const anyVar: any;
declare const nestedAny: { prop: any };

anyVar();
anyVar.a.b();

nestedAny.prop();
nestedAny.prop['a']();

new anyVar();
new nestedAny.prop();

anyVar`foo`;
nestedAny.prop`foo`;
```

<<<<<<< HEAD
### âœ… Correct

```ts
declare const typedVar: () => void;
declare const typedNested: { prop: { a: () => void } };

typedVar();
typedNested.prop.a();
=======
Examples of **correct** code for this rule:

```ts
declare const properlyTyped: { prop: { a: () => void } };

nestedAny.prop.a();
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

(() => {})();

new Map();

String.raw`foo`;
```

<<<<<<< HEAD
## Related To

- [`no-explicit-any`](./no-explicit-any.md)
=======
## Related to

- [`no-explicit-any`](./no-explicit-any.md)
- TSLint: [`no-unsafe-any`](https://palantir.github.io/tslint/rules/no-unsafe-any/)
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
