<<<<<<< HEAD
---
description: 'Disallow the use of variables before they are defined.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/no-use-before-define** for documentation.

## Examples
=======
# Disallow the use of variables before they are defined (`no-use-before-define`)

## PLEASE READ THIS ISSUE BEFORE USING THIS RULE [#1856](https://github.com/typescript-eslint/typescript-eslint/issues/1856)

## Rule Details
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

This rule extends the base [`eslint/no-use-before-define`](https://eslint.org/docs/rules/no-use-before-define) rule.
It adds support for `type`, `interface` and `enum` declarations.

<<<<<<< HEAD
## Options

This rule adds the following options:

```ts
interface Options extends BaseNoUseBeforeDefineOptions {
  enums?: boolean;
  typedefs?: boolean;
  ignoreTypeReferences?: boolean;
}

const defaultOptions: Options = {
  ...baseNoUseBeforeDefineDefaultOptions,
  enums: true,
  typedefs: true,
  ignoreTypeReferences: true,
=======
## How to use

```jsonc
{
  // note you must disable the base rule as it can report incorrect errors
  "no-use-before-define": "off",
  "@typescript-eslint/no-use-before-define": ["error"]
}
```

## Options

See [`eslint/no-use-before-define` options](https://eslint.org/docs/rules/no-use-before-define#options).
This rule adds the following options:

```ts
interface Options extends BaseNoMagicNumbersOptions {
  enums?: boolean;
  typedefs?: boolean;
}

const defaultOptions: Options = {
  ...baseNoMagicNumbersDefaultOptions,
  enums: true,
  typedefs: true,
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
};
```

### `enums`

<<<<<<< HEAD
If this is `true`, this rule warns every reference to a enum before the enum declaration.
If this is `false`, this rule will ignore references to enums, when the reference is in a child scope.

Examples of code for the `{ "enums": true }` option:

<!--tabs-->

#### âŒ Incorrect
=======
The flag which shows whether or not this rule checks enum declarations of upper scopes.
If this is `true`, this rule warns every reference to a enum before the enum declaration.
Otherwise, ignores those references.

Examples of **incorrect** code for the `{ "enums": true }` option:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
/*eslint no-use-before-define: ["error", { "enums": true }]*/

<<<<<<< HEAD
const x = Foo.FOO;

enum Foo {
  FOO,
}
```

#### âœ… Correct
=======
function foo() {
  return Foo.FOO;
}

class Test {
  foo() {
    return Foo.FOO;
  }
}

enum Foo {
  FOO,
  BAR,
}
```

Examples of **correct** code for the `{ "enums": false }` option:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
/*eslint no-use-before-define: ["error", { "enums": false }]*/

function foo() {
  return Foo.FOO;
}

enum Foo {
  FOO,
}
```

### `typedefs`

<<<<<<< HEAD
If this is `true`, this rule warns every reference to a type before the type declaration.
If this is `false`, this rule will ignore references to types.
=======
The flag which shows whether or not this rule checks type declarations.
If this is `true`, this rule warns every reference to a type before the type declaration.
Otherwise, ignores those references.
Type declarations are hoisted, so it's safe.
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

Examples of **correct** code for the `{ "typedefs": false }` option:

```ts
/*eslint no-use-before-define: ["error", { "typedefs": false }]*/

let myVar: StringOrNumber;
type StringOrNumber = string | number;
```

<<<<<<< HEAD
### `ignoreTypeReferences`

If this is `true`, this rule ignores all type references, such as in type annotations and assertions.
If this is `false`, this will will check all type references.

Examples of **correct** code for the `{ "ignoreTypeReferences": true }` option:

```ts
/*eslint no-use-before-define: ["error", { "ignoreTypeReferences": true }]*/

let var1: StringOrNumber;
type StringOrNumber = string | number;

let var2: Enum;
enum Enum {}
```
=======
Copied from [the original ESLint rule docs](https://github.com/eslint/eslint/blob/a113cd3/docs/rules/no-use-before-define.md)
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
