<<<<<<< HEAD
---
description: 'Disallow type arguments that are equal to the default.'
---

> üõë This file is source code, not the primary documentation location! üõë
>
> See **https://typescript-eslint.io/rules/no-unnecessary-type-arguments** for documentation.
=======
# Enforces that type arguments will not be used if not required (`no-unnecessary-type-arguments`)

Warns if an explicitly specified type argument is the default for that type parameter.

## Rule Details
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

Type parameters in TypeScript may specify a default value.
For example:

```ts
<<<<<<< HEAD
function f<T = number>(...) {...}
```

It is redundant to provide an explicit type parameter equal to that default: e.g. calling `f<number>(...)`.
This rule reports when an explicitly specified type argument is the default for that type parameter.

## Examples

<!--tabs-->

### ‚ùå Incorrect
=======
function f<T = number>() {}
```

It is redundant to provide an explicit type parameter equal to that default.

Examples of **incorrect** code for this rule:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
function f<T = number>() {}
f<number>();
<<<<<<< HEAD
```

```ts
function g<T = number, U = string>() {}
g<string, string>();
```

```ts
class C<T = number> {}
new C<number>();

class D extends C<number> {}
```

```ts
=======

function g<T = number, U = string>() {}
g<string, string>();

class C<T = number> {}
function h(c: C<number>) {}
new C<number>();
class D extends C<number> {}

>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
interface I<T = number> {}
class Impl implements I<number> {}
```

<<<<<<< HEAD
### ‚úÖ Correct

```ts
function f<T = number>() {}
f();
f<string>();
```

```ts
function g<T = number, U = string>() {}
g<string>();
g<number, number>();
```

```ts
class C<T = number> {}
new C();
new C<string>();

class D extends C {}
class D extends C<string> {}
```

```ts
interface I<T = number> {}
class Impl implements I<string> {}
```
=======
Examples of **correct** code for this rule:

```ts
function f<T = number>() {}
f<string>();

function g<T = number, U = string>() {}
g<number, number>();

class C<T = number> {}
new C<string>();
class D extends C<string> {}

interface I<T = number> {}
class Impl implements I<string> {}
```

## Related to

- TSLint: [use-default-type-parameter](https://palantir.github.io/tslint/rules/use-default-type-parameter)
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
