<<<<<<< HEAD
---
description: 'Enforce using function types instead of interfaces with call signatures.'
---

> ðŸ›‘ This file is source code, not the primary documentation location! ðŸ›‘
>
> See **https://typescript-eslint.io/rules/prefer-function-type** for documentation.

TypeScript allows for two common ways to declare a type for a function:

- Function type: `() => string`
- Object type with a signature: `{ (): string }`

The function type form is generally preferred when possible for being more succinct.

This rule suggests using a function type instead of an interface or object type literal with a single call signature.

## Examples

<!--tabs-->

### âŒ Incorrect

```ts
interface Example {
=======
# Use function types instead of interfaces with call signatures (`prefer-function-type`)

## Rule Details

This rule suggests using a function type instead of an interface or object type literal with a single call signature.

Examples of **incorrect** code for this rule:

```ts
interface Foo {
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
  (): string;
}
```

```ts
<<<<<<< HEAD
function foo(example: { (): number }): number {
  return example();
}
```

```ts
interface ReturnsSelf {
  // returns the function itself, not the `this` argument.
  (arg: string): this;
}
```

### âœ… Correct

```ts
type Example = () => string;
```

```ts
function foo(example: () => number): number {
=======
function foo(bar: { (): number }): number {
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
  return bar();
}
```

```ts
<<<<<<< HEAD
// returns the function itself, not the `this` argument.
type ReturnsSelf = (arg: string) => ReturnsSelf;
=======
interface Foo extends Function {
  (): void;
}
```

Examples of **correct** code for this rule:

```ts
interface Foo {
  (): void;
  bar: number;
}
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
```

```ts
function foo(bar: { (): string; baz: number }): string {
  return bar();
}
```

```ts
interface Foo {
  bar: string;
}
interface Bar extends Foo {
  (): void;
}
```

<<<<<<< HEAD
```ts
// multiple call signatures (overloads) is allowed:
interface Overloaded {
  (data: string): number;
  (id: number): string;
}
// this is equivelent to Overloaded interface.
type Intersection = ((data: string) => number) & ((id: number) => string);
```

=======
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
## When Not To Use It

If you specifically want to use an interface or type literal with a single call signature for stylistic reasons, you can disable this rule.

<<<<<<< HEAD
This rule has a known edge case of sometimes triggering on global augmentations such as `interface Function`.
These edge cases are rare and often symptomatic of odd code.
We recommend you use an [inline ESLint disable comment](https://eslint.org/docs/latest/use/configure/rules#using-configuration-comments-1).
See [#454](https://github.com/typescript-eslint/typescript-eslint/issues/454) for details.
=======
## Further Reading

- TSLint: [`callable-types`](https://palantir.github.io/tslint/rules/callable-types/)
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
