<<<<<<< HEAD
---
description: 'Enforce the use of `as const` over literal type.'
---

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/prefer-as-const** for documentation.

There are two common ways to tell TypeScript that a literal value should be interpreted as its literal type (e.g. `2`) rather than general primitive type (e.g. `number`);

- `as const`: telling TypeScript to infer the literal type automatically
- `as` with the literal type: explicitly telling the literal type to TypeScript

`as const` is generally preferred, as it doesn't require re-typing the literal value.
This rule reports when an `as` with an explicit literal type can be replaced with an `as const`.

## Examples

<!--tabs-->

### ❌ Incorrect
=======
# Prefer usage of `as const` over literal type (`prefer-as-const`)

This rule recommends usage of `const` assertion when type primitive value is equal to type.

## Rule Details

Examples of **incorrect** code for this rule:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
let bar: 2 = 2;
let foo = <'bar'>'bar';
let foo = { bar: 'baz' as 'baz' };
```

<<<<<<< HEAD
### ✅ Correct
=======
Examples of **correct** code for this rule:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
let foo = 'bar';
let foo = 'bar' as const;
let foo: 'bar' = 'bar' as const;
let bar = 'bar' as string;
let foo = <string>'bar';
let foo = { bar: 'baz' };
```

<<<<<<< HEAD
<!--/tabs-->

=======
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
## When Not To Use It

If you are using TypeScript < 3.4
