<<<<<<< HEAD
---
description: 'Enforce using `@ts-expect-error` over `@ts-ignore`.'
---

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/prefer-ts-expect-error** for documentation.

TypeScript allows you to suppress all errors on a line by placing a comment starting with `@ts-ignore` or `@ts-expect-error` immediately before the erroring line.
The two directives work the same, except `@ts-expect-error` causes a type error if placed before a line that's not erroring in the first place.
=======
# Recommends using `// @ts-expect-error` over `// @ts-ignore` (`prefer-ts-expect-error`)

TypeScript allows you to suppress all errors on a line by placing a single-line comment starting with `@ts-ignore` immediately before the erroring line.
While powerful, there is no way to know if a `@ts-ignore` is actually suppressing an error without manually investigating what happens when the `@ts-ignore` is removed.
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

This means its easy for `@ts-ignore`s to be forgotten about, and remain in code even after the error they were suppressing is fixed.
This is dangerous, as if a new error arises on that line it'll be suppressed by the forgotten about `@ts-ignore`, and so be missed.

<<<<<<< HEAD
## Examples

This rule reports any usage of `@ts-ignore`, including a fixer to replace with `@ts-expect-error`.

<!--tabs-->

### ❌ Incorrect
=======
To address this, TS3.9 ships with a new single-line comment directive: `// @ts-expect-error`.

This directive operates in the same manner as `@ts-ignore`, but will error if the line it's meant to be suppressing doesn't actually contain an error, making it a lot safer.

## Rule Details

This rule looks for usages of `@ts-ignore`, and flags them to be replaced with `@ts-expect-error`.

Examples of **incorrect** code for this rule:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
// @ts-ignore
const str: string = 1;

<<<<<<< HEAD
/**
 * Explaining comment
 *
 * @ts-ignore */
const multiLine: number = 'value';

/** @ts-ignore */
const block: string = 1;

=======
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
const isOptionEnabled = (key: string): boolean => {
  // @ts-ignore: if key isn't in globalOptions it'll be undefined which is false
  return !!globalOptions[key];
};
```

<<<<<<< HEAD
### ✅ Correct
=======
Examples of **correct** code for this rule:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
// @ts-expect-error
const str: string = 1;

<<<<<<< HEAD
/**
 * Explaining comment
 *
 * @ts-expect-error */
const multiLine: number = 'value';

/** @ts-expect-error */
const block: string = 1;

=======
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
const isOptionEnabled = (key: string): boolean => {
  // @ts-expect-error: if key isn't in globalOptions it'll be undefined which is false
  return !!globalOptions[key];
};
```

## When Not To Use It

<<<<<<< HEAD
If you are compiling against multiple versions of TypeScript and using `@ts-ignore` to ignore version-specific type errors, this rule might get in your way.
=======
If you are not using TypeScript 3.9 (or greater), then you will not be able to use this rule, as the directive is not supported
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

## Further Reading

- [Original Implementing PR](https://github.com/microsoft/TypeScript/pull/36014)
