<<<<<<< HEAD
---
description: "Require private members to be marked as `readonly` if they're never modified outside of the constructor."
---

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/prefer-readonly** for documentation.
=======
# Requires that private members are marked as `readonly` if they're never modified outside of the constructor (`prefer-readonly`)

This rule enforces that private members are marked as `readonly` if they're never modified outside of the constructor.

## Rule Details
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

Member variables with the privacy `private` are never permitted to be modified outside of their declaring class.
If that class never modifies their value, they may safely be marked as `readonly`.

<<<<<<< HEAD
This rule reports on private members are marked as `readonly` if they're never modified outside of the constructor.

## Examples

<!--tabs-->

### ❌ Incorrect
=======
Examples of **incorrect** code for this rule:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Container {
  // These member variables could be marked as readonly
  private neverModifiedMember = true;
  private onlyModifiedInConstructor: number;

  public constructor(
    onlyModifiedInConstructor: number,
    // Private parameter properties can also be marked as readonly
    private neverModifiedParameter: string,
  ) {
    this.onlyModifiedInConstructor = onlyModifiedInConstructor;
  }
}
```

<<<<<<< HEAD
### ✅ Correct
=======
Examples of **correct** code for this rule:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Container {
  // Public members might be modified externally
  public publicMember: boolean;

  // Protected members might be modified by child classes
  protected protectedMember: number;

  // This is modified later on by the class
  private modifiedLater = 'unchanged';

  public mutate() {
    this.modifiedLater = 'mutated';
  }
}
```

## Options

<<<<<<< HEAD
=======
This rule, in its default state, does not require any argument.

>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
### `onlyInlineLambdas`

You may pass `"onlyInlineLambdas": true` as a rule option within an object to restrict checking only to members immediately assigned a lambda value.

```jsonc
{
  "@typescript-eslint/prefer-readonly": ["error", { "onlyInlineLambdas": true }]
}
```

<<<<<<< HEAD
Example of code for the `{ "onlyInlineLambdas": true }` options:

<!--tabs-->

#### ❌ Incorrect
=======
Example of **correct** code for the `{ "onlyInlineLambdas": true }` options:

```ts
class Container {
  private neverModifiedPrivate = 'unchanged';
}
```

Example of **incorrect** code for the `{ "onlyInlineLambdas": true }` options:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
class Container {
  private onClick = () => {
    /* ... */
  };
}
```

<<<<<<< HEAD
#### ✅ Correct

```ts
class Container {
  private neverModifiedPrivate = 'unchanged';
}
```
=======
## Related to

- TSLint: ['prefer-readonly'](https://palantir.github.io/tslint/rules/prefer-readonly)
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
