<<<<<<< HEAD
---
description: 'Enforce consistent returning of awaited values.'
---

> 🛑 This file is source code, not the primary documentation location! 🛑
>
> See **https://typescript-eslint.io/rules/return-await** for documentation.

Returning an awaited promise can make sense for better stack trace information as well as for consistent error handling (returned promises will not be caught in an async function try/catch).

## Examples
=======
# Enforces consistent returning of awaited values (`return-await`)

Returning an awaited promise can make sense for better stack trace information as well as for consistent error handling (returned promises will not be caught in an async function try/catch).

## Rule Details
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

This rule builds on top of the [`eslint/no-return-await`](https://eslint.org/docs/rules/no-return-await) rule.
It expands upon the base rule to add support for optionally requiring `return await` in certain cases.

<<<<<<< HEAD
=======
## How to use

```jsonc
{
  // note you must disable the base rule as it can report incorrect errors
  "no-return-await": "off",
  "@typescript-eslint/return-await": "error"
}
```

>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
## Options

```ts
type Options = 'in-try-catch' | 'always' | 'never';

const defaultOptions: Options = 'in-try-catch';
```

### `in-try-catch`

Requires that a returned promise must be `await`ed in `try-catch-finally` blocks, and disallows it elsewhere.
<<<<<<< HEAD
Specifically:

- if you `return` a promise within a `try`, then it must be `await`ed.
- if you `return` a promise within a `catch`, and there **_is no_** `finally`, then it **_must not_** be `await`ed.
- if you `return` a promise within a `catch`, and there **_is a_** `finally`, then it **_must_** be `await`ed.
- if you `return` a promise within a `finally`, then it **_must not_** be `await`ed.

Examples of code with `in-try-catch`:

<!--tabs-->

#### ❌ Incorrect
=======

Examples of **incorrect** code with `in-try-catch`:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
async function invalidInTryCatch1() {
  try {
    return Promise.resolve('try');
  } catch (e) {}
}

async function invalidInTryCatch2() {
<<<<<<< HEAD
  try {
    throw new Error('error');
  } catch (e) {
    return await Promise.resolve('catch');
  }
}

async function invalidInTryCatch3() {
  try {
    throw new Error('error');
  } catch (e) {
    return Promise.resolve('catch');
  } finally {
    console.log('cleanup');
  }
}

async function invalidInTryCatch4() {
  try {
    throw new Error('error');
  } catch (e) {
    throw new Error('error2');
  } finally {
    return await Promise.resolve('finally');
  }
}

async function invalidInTryCatch5() {
  return await Promise.resolve('try');
}

async function invalidInTryCatch6() {
=======
  return await Promise.resolve('try');
}

async function invalidInTryCatch3() {
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
  return await 'value';
}
```

<<<<<<< HEAD
#### ✅ Correct
=======
Examples of **correct** code with `in-try-catch`:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
async function validInTryCatch1() {
  try {
    return await Promise.resolve('try');
  } catch (e) {}
}

async function validInTryCatch2() {
<<<<<<< HEAD
  try {
    throw new Error('error');
  } catch (e) {
    return Promise.resolve('catch');
  }
}

async function validInTryCatch3() {
  try {
    throw new Error('error');
  } catch (e) {
    return await Promise.resolve('catch');
  } finally {
    console.log('cleanup');
  }
}

async function validInTryCatch4() {
  try {
    throw new Error('error');
  } catch (e) {
    throw new Error('error2');
  } finally {
    return Promise.resolve('finally');
  }
}

async function validInTryCatch5() {
  return Promise.resolve('try');
}

async function validInTryCatch6() {
=======
  return Promise.resolve('try');
}

async function validInTryCatch3() {
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa
  return 'value';
}
```

### `always`

Requires that all returned promises are `await`ed.

<<<<<<< HEAD
Examples of code with `always`:

<!--tabs-->

#### ❌ Incorrect
=======
Examples of **incorrect** code with `always`:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
async function invalidAlways1() {
  try {
    return Promise.resolve('try');
  } catch (e) {}
}

async function invalidAlways2() {
  return Promise.resolve('try');
}

async function invalidAlways3() {
  return await 'value';
}
```

<<<<<<< HEAD
#### ✅ Correct
=======
Examples of **correct** code with `always`:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
async function validAlways1() {
  try {
    return await Promise.resolve('try');
  } catch (e) {}
}

async function validAlways2() {
  return await Promise.resolve('try');
}

async function validAlways3() {
  return 'value';
}
```

### `never`

Disallows all `await`ing any returned promises.

<<<<<<< HEAD
Examples of code with `never`:

<!--tabs-->

#### ❌ Incorrect
=======
Examples of **incorrect** code with `never`:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
async function invalidNever1() {
  try {
    return await Promise.resolve('try');
  } catch (e) {}
}

async function invalidNever2() {
  return await Promise.resolve('try');
}

async function invalidNever3() {
  return await 'value';
}
```

<<<<<<< HEAD
#### ✅ Correct
=======
Examples of **correct** code with `never`:
>>>>>>> 8aded176e2bf64ab331edd712b5a251ef6940ffa

```ts
async function validNever1() {
  try {
    return Promise.resolve('try');
  } catch (e) {}
}

async function validNever2() {
  return Promise.resolve('try');
}

async function validNever3() {
  return 'value';
}
```
